### PR TITLE
Traffic Ops Go client cleanup

### PR DESCRIPTION
**Pull request by ocket8888**
## What does this PR (Pull Request) do?
- [x] This PR is not related to any Issue

This PR primarily removes client methods that were marked deprecated in the old, un-versioned `github.com/apache/trafficcontrol/traffic_ops/client` package. It then removes the now-unnecessary qualifiers from method names. For example, if the client currently has methods:

```go
func (to *Session) GetSomethingWithHdr(header http.Header) (ResponseType, ReqInf, error) {
    // Actually do something
}

func (to *Session) GetSomething() (ResponseType, ReqInf, error) {
    return to.GetSomethingWithHdr(nil)
}
```

... this PR would replace it with:

```go
func (to *Session) GetSomething(header http.Header) (ResponseType, ReqInf, error) {
    // Actually do something
}
```

This PR also collapses duplicate methods, like turning: 
```go
func (to *Session) GetSomethingWithHdr(header http.Header) (ResponseType, ReqInf, error) {
    return to.GGetSomethingWithQueryParamsAndHdr(nil, header)
}

func (to *Session) GetSomethingWithQueryParamsAndHdr(qparams url.Values, header http.Header) (ResponseType, ReqInf, error) {
    // Actually do something
}
```

into:

```go
func (to *Session) GetSomething(qparams url.Values, header http.Header) (ResponseType, ReqInf, error) {
    // Actually do something
}
```

Furthermore, there were roughly four different ways we were passing arbitrary query string parameters into methods

* Using actual `net/url.Values` structures
* Using a pointer to a `net/url.Values` structure
* Just a raw `map[string]string`
* A literal string that is concatenated onto the end of a request path

These have been consolidated to use `net/url.Values`, (at least I hope I got them all).

Because of an issue where a client method was duplicated - specifically `GetParameters` takes query parameters for filtering but `GetParametersByValue` existed and filtered client-side by iterating the response - because it provided functionality not offered by the TO API itself, this PR also adds support for the "value" query parameter to `/parameters`. This was a one-line code change in TO (docs changes make it a bit longer), so I'm leaving it in this open PR, but I also opened #5508 which would do that separately and if it's deemed too big to squeeze into this PR then I'll put this in draft until that's merged.

This PR also adds a bunch of GoDoc comments, removes the `/deliveryservice_user` methods/file since that isn't in APIv4, enforces proper ID types (a bunch of methods took strings for an `int` field which essentially just caused a lot of `strconv.Itoa` in the places where they were called), and changes most functions that took/returned ATC object type _pointers_ to just use the types, to avoid requests having side effects on the caller's structures and to avoid the caller having to do a lot of `nil` checks that ought to be unnecessary.

Finally, because the CDN-in-a-Box enroller links against the v4 API client, some changes to the client also necessitate changes to the enroller.

## Which Traffic Control components are affected by this PR?
- CDN-in-a-Box
- Documentation
- Traffic Ops Client (Go)
- Traffic Ops

## What is the best way to verify this PR?
Ensure the Go client/API integration tests all still pass, CDN-in-a-Box can still build and run, docs build without errors or warnings, and read through the Go client code to ensure I didn't miss any deprecated/duplicated methods or uses of non-`net/url.Values` query parameters.

## The following criteria are ALL met by this PR
- [x] This PR includes tests
- [x] This PR includes documentation
- [x] This PR includes an update to CHANGELOG.md
- [x] This PR includes any and all required license headers
- [x] This PR **DOES NOT FIX A SERIOUS SECURITY VULNERABILITY**

### CODE REVIEW

**Comment by rawlinp** [comment]
I would probably hold off on this one until https://github.com/apache/trafficcontrol/pull/5430 is completed/merged. That PR is actually needed for 5.1.

Also, wouldn't this be a good PR to implement the TO Go client best practices discussed on the mailing list? I wasn't sure if we really wanted to make all these breaking signature changes. But if we are, it might also be a good time to move resources into their own "modules" a la `Session.DeliveryServices.Get(...)`, `Session.Servers.Update(...)`?

**Comment by ocket8888** [comment]
> I would probably hold off on this one until #5430 is completed/merged. That PR is actually needed for 5.1.

Yeah, sure thing.

> wouldn't this be a good PR to implement the TO Go client best practices discussed on the mailing list?

I mean, I could. I don't have a problem with doing the work, I just thought at ~4.2k lines changed this PR was big enough. Basically this is the first step in accomplishing that, just cutting down the client methods to the ones that should exist and implementing query parameters one way to make the transition to RequestOptions easier.

>  wasn't sure if we really wanted to make all these breaking signature changes

Because `traffic_ops/client` is deprecated/soon to be removed and people no longer import the same path for each, continually upgrading API client version we can absolutely just shatter the call signatures. It's not breaking anyone because they're explicitly import `traffic_ops/v3-client` which is totally unaffected. That's one of the real big advantages to having a separate client for each version.

> But if we are, it might also be a good time to move resources into their own "modules" a la `Session.DeliveryServices.Get(...)`, `Session.Servers.Update(...)`?

Maybe, "time" being a sort of flexible concept in this context. Between major revisions in an unreleased API version is a good "time", but that sounds like a huge amount of work that I wouldn't want to shove into this PR. I'm actually not entirely certain what that would look like code-wise and whether or not I'm sold on the idea.

**Comment by ocket8888** [comment]
Draft pending #5430

**Comment by rawlinp** [comment]
> It's not breaking anyone because they're explicitly import traffic_ops/v3-client which is totally unaffected.

Sure, it's not breaking them without their choice, but it makes it a lot harder to transition to the new clients when required. It would be better to pay the price all at once IMO than to get broken multiple times when transitioning to newer clients.

**Comment by rawlinp** [comment]
That said, I do feel you on the PR size issue. It doesn't necessarily need to all be done in one PR, but it would be great to complete all the breaking changes we want in the span of a single release. That way you only have to pay the transition price once. Then going forward, users should mostly just be able to change their imports from `vX-client` to `vY-client`, plus any route-specific breaking changes that were made.

**Comment by ocket8888** [comment]
Yeah, that's the plan. I'm already working on the next round of changes on top of this; wanna move fast.

**Comment by rawlinp** [comment]
So https://github.com/apache/trafficcontrol/pull/5430 didn't make the changes to the v4-client, only the v3-client. So we should probably try to complete https://github.com/apache/trafficcontrol/issues/5518 before we merge this. The more out of sync the two versions get, the harder/more tedious it will be to copy that functionality into the v4-client.

**Comment by ocket8888** [comment]
Draft pending #5518

File: traffic_ops/v4-client/cdn.go
Comment (srijeet0406) [inline traffic_ops/v4-client/cdn.go:40]: I think I've made this comment before, but this could just be `UpdateCDN`

File: traffic_ops/v4-client/cdn.go
Comment (srijeet0406) [inline traffic_ops/v4-client/cdn.go:71]: This could just be `DeleteCDN`

File: traffic_ops/v4-client/cdnfederations.go
Comment (srijeet0406) [inline traffic_ops/v4-client/cdnfederations.go:70]: Should this just be `DeleteCDNFederation`?

File: traffic_ops/v4-client/coordinate.go
Comment (srijeet0406) [inline traffic_ops/v4-client/coordinate.go:70]: Same comment here

File: traffic_ops/v4-client/crconfig.go
Comment (srijeet0406) [inline traffic_ops/v4-client/crconfig.go:82]: I think we should pass in the header param here as well.

File: traffic_ops/v4-client/deliveryservice.go
Comment (srijeet0406) [inline traffic_ops/v4-client/deliveryservice.go:118]: I think I mentioned this before, but could you make `header` the last argument?

File: traffic_ops/v4-client/server_update_status.go
Comment (srijeet0406) [inline traffic_ops/v4-client/server_update_status.go:30]: You could pass in the `header` here as well.

**Review by srijeet0406 — COMMENTED** [review]
Code looks good overall. I had some minor comments about `nil` checking and dealing with the `header` param in some of the functions.

File: traffic_ops/testing/api/v4/federation_resolvers_test.go
Comment (srijeet0406) [reply traffic_ops/testing/api/v4/federation_resolvers_test.go:173]: But this could be its own separate PR if you'd like.

File: traffic_ops/testing/api/v4/cachegroups_parameters_test.go
Comment (ocket8888) [reply traffic_ops/testing/api/v4/cachegroups_parameters_test.go:51]: I'm going to follow this PR up with another that makes all the call signatures of the methods consistent. I was going to do this then, but it's easy enough to do it now if you want.

Comment (ocket8888) [reply traffic_ops/testing/api/v4/cachegroups_parameters_test.go:51]: That would include adding the HTTP header parameter where it's missing to method signatures, as well.

Comment (srijeet0406) [reply traffic_ops/testing/api/v4/cachegroups_parameters_test.go:51]: Ok cool, I'm fine either way. If you're going to address it in a follow up PR, I'd say add a comment in the PR description saying that this is part 1/2 or something like that.

File: CHANGELOG.md
Comment (ocket8888) [reply CHANGELOG.md:37]: No, must just be the result of a bad rebase

File: traffic_ops/testing/api/v4/cdnfederations_test.go
Comment (ocket8888) [reply traffic_ops/testing/api/v4/cdnfederations_test.go:126]: That should actually just not return a reference value. Changed

File: traffic_ops/testing/api/v4/cdns_test.go
Comment (ocket8888) [reply traffic_ops/testing/api/v4/cdns_test.go:59]: Yes. And `DeleteCDNByID` should likewise just be `DeleteCDN`

File: traffic_ops/testing/api/v4/federation_resolvers_test.go
Comment (ocket8888) [reply traffic_ops/testing/api/v4/federation_resolvers_test.go:173]: Yes, the plan is to make all client methods take query parameter and HTTP header arguments.

File: traffic_ops/testing/api/v4/profiles_test.go
Comment (ocket8888) [reply traffic_ops/testing/api/v4/profiles_test.go:87]: I suppose it can be removed, that's why we use git, after all

File: traffic_ops/testing/api/v4/federation_resolvers_test.go
Comment (rob05c) [reply traffic_ops/testing/api/v4/federation_resolvers_test.go:173]: Can I suggest putting both in some kind of `Opts` struct? The advantage of that, is so future arguments can be added to the end of the struct without breaking existing things. 

So we don't need to add a new function with a new name, existing calls still work, the function signature just stays e.g. `func (cl *TOClient) CreateFederationResolver(fed tc.FederationResolver, opts Opts)`

Not for this PR, just a future suggestion/proposal

File: traffic_ops/v4-client/cdn.go
Comment (ocket8888) [reply traffic_ops/v4-client/cdn.go:40]: You did

File: traffic_ops/testing/api/v4/federation_resolvers_test.go
Comment (rawlinp) [reply traffic_ops/testing/api/v4/federation_resolvers_test.go:173]: Indeed, that was what we agreed on the mailing list, and I'm sure @ocket8888 is saving that for a follow-up PR to not make this one supermassive.

File: traffic_ops/testing/api/v4/cachegroups_test.go
Comment (srijeet0406) [inline traffic_ops/testing/api/v4/cachegroups_test.go:532]: I think we should change this to fatal here. Its dereferencing `cg.TypeID` on line 551.

Comment (ocket8888) [reply traffic_ops/testing/api/v4/cachegroups_test.go]: Yes, that should've been fatal

**Review by srijeet0406 — APPROVED** [review]
Code looks good. API/ unit/ manual tests passed. All review comments have been addressed. LGTM!

**Pull request merged**
